waspc/data/Generator/templates/server/src/entities/index.ts

```diff
@@ -4,15 +4,40 @@ import {
   type {= name =},
   {=/ entities =}
 } from "@prisma/client"
-  
+
 export {
   {=# entities =}
   type {= name =},
   {=/ entities =}
 } from "@prisma/client"
 
-export type WaspEntity = 
+type Entity = 
   {=# entities =}
   | {= name =}
   {=/ entities =}
   | never
+
+export type EntityName = 
+  {=# entities =}
+  | "{= name =}"
+  {=/ entities =}
+  | never
+
+// Wasp internally uses the following types for typing entity maps in
+// operation contexts.
+//
+// We must explicitly tag all entities with their name to avoid issues with
+// structural typing. See https://github.com/wasp-lang/wasp/pull/982 for details.
+
+{=# entities =}
+export type {= internalTypeName =} = WithName<{= name =}, "{= name =}">
+{=/ entities =}
+
+export type _Entity = 
+  {=# entities =}
+  | {= internalTypeName =}
+  {=/ entities =}
+  | never
+
+type WithName<E extends Entity, Name extends EntityName> = 
+  E & { _entityName: Name }
```

Comment:
Should we be creating "internal" files for all the stuff we don't want users to import?

They'll still be able to import anything we export, but importing from `something/_internal` will probably send the right signal.

Reply:
Ideally, it would probably be better if they were somehow more hidden or differently scoped to indicate they should not use them, but since it is easier to just put them all into a single file I think it is probably fine to leave as is.

Reply:
It makes sense for the me to keep internal stuff in `_taggedEntities.ts` or something like that. It's not uncommon in JS world to have internal files prefixed with `_`.

One argument against it is: what happens when user clicks on the `_Task` type to see what's there. If the user is taken to a isolated file with only `_Task` which imports `Task` then the user needs to again click on the `Task` to see the definition. Keep `_Task` and `Task` in the same file helps. 

If we don't care that much about the second part, then I'm okay with moving this to an internal file, which will keep the entity types file cleaner.

Reply:
Did we decide if we want to hide this in an internal file?

Reply:
Yes, it's hidden now. Also, the entire `types` modules is now internal.